packages/engine/src/gltf/GLTFLoader.test.tsx

```diff
@@ -0,0 +1,757 @@
+import assert from 'assert'
+import React, { useEffect } from 'react'
+import Sinon from 'sinon'
+import { BufferGeometry, InstancedMesh, MathUtils, MeshStandardMaterial } from 'three'
+import { AssetLoaderState } from '../assets/state/AssetLoaderState'
+import { AnimationComponent } from '../avatar/components/AnimationComponent'
+import { GLTFComponent } from './GLTFComponent'
+import { GLTFDocumentState, GLTFNode, GLTFNodeState } from './GLTFDocumentState'
+import { getNodeUUID } from './GLTFState'
+import { KHRUnlitExtensionComponent, MaterialDefinitionComponent } from './MaterialDefinitionComponent'
+import { EXTMeshGPUInstancingComponent, KHRLightsPunctualComponent, KHRPunctualLight } from './MeshExtensionComponents'
+import { afterEach, beforeEach, describe, it } from 'vitest'
+
+const CDN_URL = 'https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/main/2.0'
+const duck_gltf = CDN_URL + '/Duck/glTF/Duck.gltf'
+const draco_box = CDN_URL + '/Box/glTF-Draco/Box.gltf'
+const unlit_gltf = CDN_URL + '/UnlitTest/glTF/UnlitTest.gltf'
+const textured_gltf = CDN_URL + '/BoxTextured/glTF/BoxTextured.gltf'
+const multiple_mesh_primitives_gltf = CDN_URL + '/CesiumMilkTruck/glTF/CesiumMilkTruck.gltf'
+const morph_gltf = CDN_URL + '/AnimatedMorphCube/glTF/AnimatedMorphCube.gltf'
+const skinned_gltf = CDN_URL + '/Fox/glTF/Fox.gltf'
+const camera_gltf = CDN_URL + '/Cameras/glTF/Cameras.gltf'
+const khr_light_gltf = CDN_URL + '/LightsPunctualLamp/glTF/LightsPunctualLamp.gltf'
+const instanced_gltf = CDN_URL + '/SimpleInstancing/glTF/SimpleInstancing.gltf'
+
+const gltfCompletedIO = async (entity: Entity) => {
+  return new Promise((resolve) => {
+    const gltfComponent = getComponent(entity, GLTFComponent)
+    const wait = () => {
+      if (!gltfComponent.dependencies) setTimeout(wait, 100)
+      else resolve(null)
+    }
+    wait()
+  })
+}
+
+// Needed when the component relies on a file read before being created (ie. materials with textures)
+const componentsLoaded = async (entity: Entity, components: ComponentType<any>[], expected: number) => {
+  return new Promise((resolve) => {
+    const wait = () => {
+      const entities = getChildrenWithComponents(entity, components)
+      if (entities.length !== expected) setTimeout(wait, 100)
+      else resolve(null)
+    }
+    wait()
+  })
+}
+
+const setupEntity = () => {
+  const parent = createEntity()
+  setComponent(parent, SceneComponent)
+  setComponent(parent, EntityTreeComponent)
+  const uuid = setComponent(parent, UUIDComponent, generateEntityUUID())
+
+  getMutableState(RapierWorldState).merge({
+    [uuid]: true as any
+  })
+
+  const entity = createEntity()
+  setComponent(entity, EntityTreeComponent, { parentEntity: parent })
+  return entity
+}
+
+describe('GLTF Loader', () => {
+  beforeEach(async () => {
+    createEngine()
+  })
+
+  afterEach(() => {
+    return destroyEngine()
+  })
+
+  it('can load a mesh', () => new Promise<void>(done => {
```

Comment:
@MichaelEstes we should refactor these tests into less nested and async functions. if we have nested reactive logic for tests. we are more prone to timeouts as things get stuck, whereas as using `await act(() => rerender(<></>))` will cause all react updates to be handled  and we can immediately see what asserts fail.